signature/plugin.go

```diff
@@ -220,25 +221,42 @@ func (s *pluginSigner) generateSignatureEnvelope(ctx context.Context, desc notat
 		return nil, err
 	}
 
+	content := envContent.Payload.Content
 	var signedPayload notation.Payload
-	if err = json.Unmarshal(envContent.Payload.Content, &signedPayload); err != nil {
+	if err = json.Unmarshal(content, &signedPayload); err != nil {
 		return nil, fmt.Errorf("signed envelope payload can't be unmarshaled: %w", err)
 	}
 
-	// TODO: Verify plugin didnot add any additional top level payload attributes. https://github.com/notaryproject/notation-go/issues/80
-	if !descriptorPartialEqual(desc, signedPayload.TargetArtifact) {
+	if !isPayloadDescriptorValid(content, desc, signedPayload.TargetArtifact) {
 		return nil, errors.New("descriptor subject has changed")
 	}
 
 	return resp.SignatureEnvelope, nil
 }
 
-// descriptorPartialEqual checks if the both descriptors point to the same resource
-// and that newDesc hasn't replaced or overridden existing annotations.
-func descriptorPartialEqual(original, newDesc notation.Descriptor) bool {
-	if !original.Equal(newDesc) {
-		return false
-	}
+func isPayloadDescriptorValid(originalContent []byte, originalDesc, newDesc notation.Descriptor) bool {
+	return originalDesc.Equal(newDesc) &&
+		!isDescriptorAttributeAdded(originalContent) &&
+		descriptorAnnotationsPartialEqual(originalDesc, newDesc)
+}
+
+// isDescriptorAttributeAdded checks that the unmarshalled json content does not have any unexpected keys added
+func isDescriptorAttributeAdded(content []byte) bool {
+	var targetArtifactMap map[string]interface{}
+	_ = json.Unmarshal(content, &targetArtifactMap)
```

Comment:
We shouldn't ignore error.

Reply:
Error is not possible because we already successfully unmarshalled at this point

Reply:
Can you please add that comment to the code?